test/test.js

```diff
@@ -33,49 +46,178 @@ function generateTestObjects(s3Client, bucket, amount, callback) {
+      endpoint: util.format('http://%s:%d', hostname, port),
+      sslEnabled: false,
+      s3ForcePathStyle: true
+    });
+    s3Client.createBucket({ Bucket: bucket }, function (err) {
+      if (err) return done(err);
+      generateTestObjects(s3Client, bucket, 10, function (err) {
+        if (err) return done(err);
+        s3rver.close(function (err) {
+          if (err) return done(err);
+          const exists = fs.existsSync(directory);
+          should(exists).equal(true);
+          const files = fs.readdirSync(directory);
+          should(files.length).equal(1);
+          done();
+        });
+      });
+    });
+  });
+});
+
+it('Does not clean up after close if the removeBucketsOnClose setting is not set', function (done) {
+  resetTmpDir();
+  const bucket = 'foobars';
+
+  const s3rver = new S3rver({
+    port: 4569,
+    hostname: 'localhost',
+    silent: true
+  }).run(function (err, hostname, port, directory) {
+    if (err) return done(err);
+    const s3Client = new AWS.S3({
+      accessKeyId: '123',
+      secretAccessKey: 'abc',
+      endpoint: util.format('http://%s:%d', hostname, port),
+      sslEnabled: false,
+      s3ForcePathStyle: true
+    });
+    s3Client.createBucket({ Bucket: bucket }, function (err) {
+      if (err) return done(err);
+
+      generateTestObjects(s3Client, bucket, 10, function (err) {
+        if (err) return done(err);
+        s3rver.close(function (err) {
+          if (err) return done(err);
+          const exists = fs.existsSync(directory);
+          should(exists).equal(true);
+          const files = fs.readdirSync(directory);
+          should(files.length).equal(1);
+          done();
+        });
+      });
+    });
+  });
+});
+
+it('Can delete a bucket that is empty after some key that includes a directory has been deleted', function (done) {
+  resetTmpDir();
+  const bucket = 'foobars';
+
+  const s3rver = new S3rver({
+    port: 4569,
+    hostname: 'localhost',
+    silent: true
+  }).run(function (err, hostname, port, directory) {
+    if (err) return done(err);
+    const s3Client = new AWS.S3({
+      accessKeyId: '123',
+      secretAccessKey: 'abc',
+      endpoint: util.format('http://%s:%d', hostname, port),
+      sslEnabled: false,
+      s3ForcePathStyle: true
+    });
+    s3Client.createBucket({ Bucket: bucket }).promise()
+      .then(() => s3Client.putObject({ Bucket: bucket, Key: 'foo/foo.txt', Body: 'Hello!'}).promise())
+      .then(() => s3Client.deleteObject({ Bucket: bucket, Key: 'foo/foo.txt' }).promise())
+      .then(() => s3Client.deleteBucket({ Bucket: bucket }).promise())
+      .then(() => s3rver.close(done))
+      .catch(err => s3rver.close(() => done(err)));
+  });
+});
+
 describe('S3rver Tests', function () {
-  let s3Client;
   const buckets = ['bucket1', 'bucket2', 'bucket3', 'bucket4', 'bucket5', 'bucket6'];
   let s3rver;
+  let s3Client;
 
-  beforeEach(function (done) {
+  before('Start s3rver', function (done) {
```

Comment:
I mostly did it in an attempt to make things just a bit more efficient. The test directory is still completely reset across tests so it still shouldn't be possible to make codependent tests in that section (AFAIK we aren't storing any kind of state in memory). I can undo it since I don't think it ended up having any noticeable impact.